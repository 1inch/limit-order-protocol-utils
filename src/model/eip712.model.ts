export const ORDER_STRUCTURE = [
    { name: 'salt', type: 'uint256' },
    { name: 'maker', type: 'address' },
    { name: 'receiver', type: 'address' },
    { name: 'makerAsset', type: 'address' },
    { name: 'takerAsset', type: 'address' },
    { name: 'makingAmount', type: 'uint256' },
    { name: 'takingAmount', type: 'uint256' },
    { name: 'makerTraits', type: 'uint256' },
];

export interface EIP712TypedData {
    types: EIP712Types & { EIP712Domain: EIP712Parameter[] };
    domain: EIP712Object;
    message: EIP712Object;
    primaryType: string;
}

export interface EIP712Types {
    [key: string]: EIP712Parameter[];
}

export interface EIP712Parameter {
    name: string;
    type: string;
}

export declare type EIP712ObjectValue = string | number | EIP712Object;

export interface EIP712Object {
    [key: string]: EIP712ObjectValue;
}

export interface MessageTypes {
    [additionalProperties: string]: MessageTypeProperty[];
    EIP712Domain: MessageTypeProperty[];
}

export interface MessageTypeProperty {
    name: string;
    type: string;
}
