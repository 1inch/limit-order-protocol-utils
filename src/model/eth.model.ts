/**
 * Eth address 160bits
 */
export type Address = string;