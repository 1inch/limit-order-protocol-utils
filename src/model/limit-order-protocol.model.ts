import {LimitOrderPredicateCallData} from '../limit-order-predicate.builder';

export enum ChainId {
    etherumMainnet = 1,
    binanceMainnet = 56,
    polygonMainnet = 137,
    optimismMainnet = 10,
    arbitrumMainnet = 42161,
    gnosisMainnet = 100,
    avalancheMainnet = 43114,
    fantomMainnet = 250,
    auroraMainnet = 1313161554,
    klaytnMainnet = 8217,
    zkSyncEraMainnet = 324,
}

export type LimitOrderSignature = string;

export type LimitOrderHash = string;

// RFQOrderData.expiresInTimestamp | RFQOrderData.id
export type RFQOrderInfo = string;

export interface LimitOrderData {
    maker: Address,
    receiver: Address,
    makerAsset: Address,
    takerAsset: Address,
    makingAmount: string,
    takingAmount: string,
    makerTraits: bigint | string,
}

export type ExtensionParamsWithCustomData = ExtensionParams & {
    customData: string;
}

export interface ExtensionParams {
    makerAssetSuffix: string;
    takerAssetSuffix: string;
    makingAmountGetter: string;
    takingAmountGetter: string;
    predicate: string;
    permit: string;
    preInteraction: string;
    postInteraction: string;
}


export interface RFQOrderData {
    // Index number of RFQ limit order. Example: 1
    id: number;
    wrapEth?: boolean;
    // Timestamp when the RFQ limit order will expire (seconds). Example: 1623166102
    expiresInTimestamp: number;
    makerAssetAddress: string;
    takerAssetAddress: string;
    makingAmount: string;
    takingAmount: string;
    makerAddress: string;
    /**
     * Formerly takerAddress
     */
    allowedSender?: string; // Optional, by default = ZERO_ADDRESS
}

export type Address = string;

export type MakerTraits = string;

/**
 * Compatible with EIP712Object
 */
export type LimitOrder = {
    salt: string;
    maker: Address; // maker address
    receiver: Address;
    makerAsset: Address; // maker asset address
    takerAsset: Address; // taker asset address
    makingAmount: string;
    takingAmount: string;
    makerTraits: MakerTraits;
}

export type LimitOrderWithExtension = {
    order: LimitOrder;
    extension: string;
}

/**
 * Partial from LimitOrder
 */
export type LimitOrderInteractions = {
    offsets: bigint;
    interactions: string;
}

/**
 * uint40
 */
export type Nonce = number | bigint;

/**
 * seconds unit40
 */
export type PredicateTimestamp = number | bigint;

export const InteractionsFields = {
    makerAssetData: 0,
    takerAssetData: 1,
    getMakingAmount: 2,
    getTakingAmount: 3,
    predicate: 4,
    permit: 5,
    preInteraction: 6,
    postInteraction: 7,
// cuz enum has numeric keys also
} as const;

export type InteractionName = keyof typeof InteractionsFields;

export type Interactions = {
    [key in InteractionName]: string;
};

export interface RFQOrder {
    info: RFQOrderInfo;
    makerAsset: string;
    takerAsset: string;
    maker: string;
    allowedSender: string;
    makingAmount: string;
    takingAmount: string;
}

export enum LimitOrderProtocolMethods {
    getMakingAmount = 'getMakingAmount',
    getTakingAmount = 'getTakingAmount',
    arbitraryStaticCall = 'arbitraryStaticCall',
    fillOrder = 'fillOrder',
    fillOrderToWithPermit = 'fillOrderToWithPermit',
    fillOrderRFQ = 'fillOrderRFQ',
    cancelOrder = 'cancelOrder',
    cancelOrderRFQ = 'cancelOrderRFQ',
    nonce = 'nonce',
    advanceNonce = 'advanceNonce',
    increaseNonce = 'increaseNonce',
    and = 'and',
    or = 'or',
    eq = 'eq',
    lt = 'lt',
    gt = 'gt',
    timestampBelow = 'timestampBelow',
    timestampBelowAndNonceEquals = 'timestampBelowAndNonceEquals',
    nonceEquals = 'nonceEquals',
    remaining = 'remaining',
    transferFrom = 'transferFrom',
    checkPredicate = 'checkPredicate',
    remainingsRaw = 'remainingsRaw',
    simulate = 'simulate',
}
