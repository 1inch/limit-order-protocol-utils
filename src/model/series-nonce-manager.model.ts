/**
 * 0-255 uint8
 */
export type Series = bigint;

export const LimitOrderSeries = {
    /**
     * @deprecated Don't use!
     * Preserved just to indicate which values are already takken.
     */
    P2Pv2: <Series>BigInt(1),

    P2Pv3: <Series>BigInt(2),
    // TODO use for gasless bulk cancellation
    // Gaslessv3 = 3,
} as const;

export enum SeriesNonceManagerMethods {
    nonce = 'nonce',
    advanceNonce = 'advanceNonce',
    increaseNonce = 'increaseNonce',
    nonceEquals = 'nonceEquals',
}