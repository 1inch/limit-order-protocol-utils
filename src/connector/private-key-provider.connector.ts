import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
