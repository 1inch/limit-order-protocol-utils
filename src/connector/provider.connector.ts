import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
