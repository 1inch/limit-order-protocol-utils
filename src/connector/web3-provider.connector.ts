import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
