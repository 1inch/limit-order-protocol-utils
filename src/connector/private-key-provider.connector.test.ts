import Web3 from 'web3';
import {instance, mock, verify, when} from 'ts-mockito';
import {
    EIP712_DOMAIN,
    LIMIT_ORDER_PROTOCOL_V3_ABI,
} from '../limit-order-protocol.const';
import {Eth} from 'web3-eth';
import {PrivateKeyProviderConnector} from './private-key-provider.connector';
import {LimitOrderLegacy} from '../model/limit-order-protocol.model';
import {EIP712TypedData, ORDER_STRUCTURE_LEGACY} from '../model/eip712.model';

describe('PrivateKeyProviderConnector', () => {
    let web3Provider: Web3;
    let privateKeyProviderConnector: PrivateKeyProviderConnector;

    const testPrivateKey =
        'd8d1f95deb28949ea0ecc4e9a0decf89e98422c2d76ab6e5f736792a388c56c7';
    const limitOrder: LimitOrderLegacy = {
        salt: "618054093254",
        makerAsset: "0xe9e7cea3dedca5984780bafc599bd69add087d56",
        takerAsset: "0x111111111117dc0aa78b770fa6a738034120c302",
        maker: "0xfb3c7eb936cAA12B5A884d612393969A557d4307",
        receiver: "0x0000000000000000000000000000000000000000",
        allowedSender: "0x0000000000000000000000000000000000000000",
        makingAmount: "1000000000000000000",
        takingAmount: "1000000000000000000",
        offsets: "9813420589127697917471531885823684359047649055178615469676279994777600",
        // eslint-disable-next-line max-len
        interactions: "0x20b83f2d0000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000de0b6b3a76400007e2d21830000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000de0b6b3a7640000bfa7514300000000000000000000000000000000000000000000000000000068000000240000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000006863592c2b0000000000000000000000000000000000000000000000000000000063593ad9cf6fc6e3000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d43070000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      };
    const typedData: EIP712TypedData = {
        primaryType: 'Order',
        types: {
            EIP712Domain: EIP712_DOMAIN,
            Order: ORDER_STRUCTURE_LEGACY,
        },
        domain: {
            name: '1inch Aggregation Router',
            version: '5',
            chainId: 1,
            verifyingContract: '',
        },
        message: limitOrder,
    };

    beforeEach(() => {
        web3Provider = mock<Web3>();
        privateKeyProviderConnector = new PrivateKeyProviderConnector(
            testPrivateKey,
            instance(web3Provider)
        );
    });

    it('signTypedData() must sign typed data by private key', async () => {
        const walletAddress = '0xa07c1d51497fb6e66aa2329cecb86fca0a957fdb';

        const signature = await privateKeyProviderConnector.signTypedData(
            walletAddress,
            typedData
        );

        expect(signature).toMatchSnapshot();
    });

    it('contractEncodeABI() changed address from null to undefined for contract instance', async () => {
        const eth = mock<Eth>();
        class ContractMock {
            methods = {
                foo: () => ({encodeABI: () => ''}),
            };
        }

        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        when(eth.Contract).thenReturn(ContractMock as any);
        when(web3Provider.eth).thenReturn(instance(eth));

        privateKeyProviderConnector.contractEncodeABI(
            LIMIT_ORDER_PROTOCOL_V3_ABI,
            null,
            'foo',
            []
        );

        verify(eth.Contract).once();
    });
});
