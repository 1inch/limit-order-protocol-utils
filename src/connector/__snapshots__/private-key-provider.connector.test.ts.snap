// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateKeyProviderConnector signTypedData() must sign typed data by private key 1`] = `"0xae84710a5b27736e17be0cde5b02753e4bfe8b2558c371b932ce935b3fbf61976ed86fac1f969b6f8869926ab9fa74bb7d5539ce8081abc5bd02c97c7347ab3a1b"`;
